clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -2490,85 +2524,146 @@ scheduler_tmpl_t<RecordType, ReaderType>::close_schedule_segment(output_ordinal_
 }
 
 template <typename RecordType, typename ReaderType>
 typename scheduler_tmpl_t<RecordType, ReaderType>::stream_status_t
-scheduler_tmpl_t<RecordType, ReaderType>::pop_from_ready_queue(
-    output_ordinal_t for_output, input_info_t *&new_input)
+scheduler_tmpl_t<RecordType, ReaderType>::pop_from_ready_queue_hold_locks(
+    output_ordinal_t from_output, output_ordinal_t for_output, input_info_t *&new_input,
+    bool from_back)
 {
-    assert(!need_sched_lock() || sched_lock_.owned_by_cur_thread());
+    assert(!need_output_lock() ||
+           (outputs_[from_output].runqueue.lock->owned_by_cur_thread() &&
+            (from_output == for_output || for_output == INVALID_OUTPUT_ORDINAL ||
+             outputs_[for_output].runqueue.lock->owned_by_cur_thread())));
     std::set<input_info_t *> skipped;
     std::set<input_info_t *> blocked;
     input_info_t *res = nullptr;
     sched_type_t::stream_status_t status = STATUS_OK;
-    uint64_t cur_time = (num_blocked_ > 0) ? get_output_time(for_output) : 0;
-    while (!ready_priority_.empty()) {
-        if (options_.randomize_next_input) {
-            res = ready_priority_.get_random_entry();
-            ready_priority_.erase(res);
+    uint64_t cur_time = get_output_time(from_output);
+    while (!outputs_[from_output].runqueue.queue.empty()) {
+        if (from_back) {
+            res = outputs_[from_output].runqueue.queue.back();
+            outputs_[from_output].runqueue.queue.erase(res);
+        } else if (options_.randomize_next_input) {
+            res = outputs_[from_output].runqueue.queue.get_random_entry();
+            outputs_[from_output].runqueue.queue.erase(res);
         } else {
-            res = ready_priority_.top();
-            ready_priority_.pop();
+            res = outputs_[from_output].runqueue.queue.top();
+            outputs_[from_output].runqueue.queue.pop();
         }
         assert(!res->unscheduled ||
                res->blocked_time > 0); // Should be in unscheduled_priority_.
-        if (res->binding.empty() || res->binding.find(for_output) != res->binding.end()) {
+        if (res->binding.empty() || for_output == INVALID_OUTPUT_ORDINAL ||
+            res->binding.find(for_output) != res->binding.end()) {
             // For blocked inputs, as we don't have interrupts or other regular
             // control points we only check for being unblocked when an input
             // would be chosen to run.  We thus keep blocked inputs in the ready queue.
             if (res->blocked_time > 0) {
                 assert(cur_time > 0);
-                --num_blocked_;
+                --outputs_[from_output].runqueue.num_blocked;
             }
             if (res->blocked_time > 0 &&
+                // XXX i#6966: We have seen wall-clock time go backward, which
+                // underflows here and always matches this condition currently.
                 cur_time - res->blocked_start_time < res->blocked_time) {
```

Comment:
The comment is misworded: what is "always matched" is unblocking the input, the opposite of what should happen.  I'm deliberately not fixing it here to avoid mixing too many behavior changes in one PR: this is already a massive change in behavior and I want to isolate this; I have this fix on on my list and will do it afterward (I did a number of other fixes beforehand, but they were more in the critical path while this seems incidental and can wait).  I did update the comment to be clear.